changelog/512.txt

```diff
@@ -0,0 +1,3 @@
+```release-note:bug
```

Comment:
Could you describe the effect of the fix rather than what you changed in which function? Something in the likes of 

> fixed panic on invalid AppRole login request

Reply:
I'm not quite sure the extent of the implications of this, hence why I worded it like I did :-) 

I _think_, theoretically, this could occur on any login method. However, the `userpass` method doesn't work as `username` is forced to be a string by virtue of it being part of the URL. CertAuth likewise doesn't use any request data. [Radius is maybe applicable](https://openbao.org/api-docs/auth/radius/#login) as `username` in the URL is optional.... so it is probably on a case-by-case basis, but definitely broader than AppRole.

However, on the same type of request that panics AppRole, applied to Radius, I don't get a panic:

```
$ curl     --request POST     --header "Content-Type: application/json"     --data '{"username": {}, "password": ""}'     'http://localhost:8200/v1/auth/radius/login'
{"errors":["Field validation failed: error converting input map[] for field \"username\": '' expected type 'string', got unconvertible type 'map[string]interface {}', value: 'map[]'"]}
```

(despite AppRole still panicing).

The same also holds for Kubernetes, which might theoretically also be applicable:

```
$ bao auth enable kubernetes
Success! Enabled kubernetes auth method at: kubernetes/
$ curl     --request POST     --header "Content-Type: application/json"     --data '{"jwt": {}, "role": {}}'     'http://localhost:8200/v1/auth/kubernetes/login'
{"errors":["Field validation failed: error converting input map[] for field \"jwt\": '' expected type 'string', got unconvertible type 'map[string]interface {}', value: 'map[]'"]}
```

But AppRole on this server definitely crashes:

```
$ curl     --request POST     --header "Content-Type: application/json"     --data '{"role_id": {}, "secret_id": ""}'     http://localhost:8200/v1/auth/approle/login
curl: (52) Empty reply from server
```

So I think perhaps we're at:

```
fix server panic on login requests with invalid parameter typing
```

in the assumption it is broader than just AppRole, but without having concrete confirmation?

Reply:
Ah, I understand now.

`c.isUserLocked` calls `c.getLoginUserInfoKey` which calls the affected `c.aliasNameFromLoginRequest`, but the first method is only called when user lockout is not disabled. We define lockout-able mount types in `configutil.GetSupportedUserLockoutsAuthMethods()` which currently is defined as `"userpass", "approle", "ldap"`. Both userpass and LDAP have the username in the request path, which forces it to be a string and thus match the type of the field data.

Other mounts don't panic, as this alias handler isn't called at all, and thus they return a regular error when the actual request handler is called.

Thus only AppRole is affected.